provider/datagen/src/lib.rs

```diff
@@ -206,36 +222,179 @@ impl DatagenProvider {
         TEST_PROVIDER.clone()
     }
 
-    pub(crate) fn filter_data_locales(
-        &self,
-        supported: Vec<icu_provider::DataLocale>,
-    ) -> Vec<icu_provider::DataLocale> {
-        match &self.source.options.locales {
-            options::LocaleInclude::All => supported,
-            options::LocaleInclude::Explicit(set) => supported
+    /// Gets a [`LocaleFallbacker`].
+    /// This should start returning references when we can use `std::sync::OnceLock`.
+    pub(crate) fn fallbacker(&self) -> Result<LocaleFallbacker, DataError> {
+        {
+            let lock = self.source.fallbacker_rwlock.read().unwrap();
+            match &*lock {
+                Some(fallbacker) => return Ok(fallbacker.clone()),
+                None => ()
+            }
+        }
+        let new_fallbacker = LocaleFallbacker::try_new_unstable(self)?;
+        {
+            let mut lock = self.source.fallbacker_rwlock.write().unwrap();
+            match &*lock {
+                Some(fallbacker) => return Ok(fallbacker.clone()),
+                None => {
+                    lock.replace(new_fallbacker.clone());
+                    return Ok(new_fallbacker);
+                }
+            }
+        }
+    }
+
+    /// Selects the maximal set of locales to export based on a [`DataKey`] and this datagen
+    /// provider's options bag. The locales may be later optionally deduplicated for fallback.
+    fn select_locales_for_key(&self, key: DataKey) -> Result<HashSet<DataLocale>, DataError> {
+        let fallbacker = self.fallbacker()?;
+        let fallbacker_with_config = fallbacker.for_config(LocaleFallbackConfig::from_key(key));
+
+        // Pre-process the supported locales into their normalized forms. For example, CLDR has
+        // "sr-Latn-ME", but ICU4X wants that to be "sr-ME". In order to load the data later, we
+        // use the `visit_default_script` option during fallback.
+        let supported_locales: HashSet<DataLocale> = self
+            .supported_locales_for_key(key)
+            .map_err(|e| e.with_key(key))?
+            .into_iter()
+            .map(|locale| {
+                // Perform the normalization:
+                fallbacker_with_config.fallback_for(locale).take()
+            })
+            .collect();
+
+        let locale_include = self.source.options.locales.clone();
+        let fallback_mode = self.source.options.fallback;
+
+        // Case 1: `LocaleInclude::All` simply exports all supported locales for this key.
+        // Also include special cases like Segmenter that should skip the rest of this function.
+        if matches!(locale_include, LocaleInclude::All)
+            || key == icu_segmenter::provider::DictionaryForWordOnlyAutoV1Marker::KEY
```

Comment:
I would _really_ like there to not be any key-specific logic in this module.

Reply:
https://github.com/unicode-org/icu4x/issues/3409#issuecomment-1621288544

Reply:
This is the code review discussion.

Reply:
I'll also note that this decoupling would enable the data exporter to be backed by a blob provider, enabling just-in-time data slicing that @Manishearth has wanted for some time.